tensorrt_llm/_torch/custom_ops/torch_custom_ops.py

```diff
@@ -560,6 +653,334 @@ def _(
+
+        # CuteDSL requires 1D scale tensors. If they're 2D, flatten them.
+        if act_sf.dim() == 2:
+            act_sf = act_sf.reshape(-1)
+        if weight_scale.dim() == 2:
+            weight_scale = weight_scale.reshape(-1)
+
+        return self.inner_runner(inputs=[act_fp4, weight, act_sf, weight_scale],
+                                 tactic=tactic)
+
+
+@torch.library.custom_op("trtllm::nvfp4_gemm_unified", mutates_args=())
+def nvfp4_gemm_unified(
+    act_fp4: torch.Tensor,
+    weight: torch.Tensor,
+    act_sf: torch.Tensor,
+    weight_scale: torch.Tensor,
+    alpha: torch.Tensor,
+    output_dtype: torch.dtype,
+    to_userbuffers: bool = False,
+    backend: str = "auto",
+) -> torch.Tensor:
+    """Unified NVFP4 GEMM with automatic or manual backend selection.
+
+    This function can automatically choose the best backend or force a specific backend:
+    - CUTLASS: Predefined CUTLASS configurations with auto-tuning
+    - cuBLASLt: Heuristic-based algorithms from cuBLASLt library
+    - CuteDSL: Blackwell-optimized persistent kernels (when available and inputs are valid)
+    - CUDA Core: CUDA Core implementation on SM >= 89 (Ada+), M <= 16 (explicit selection only)
+
+    The AutoTuner profiles all available backends during the first run and caches
+    the best choice for each input shape. Subsequent calls use the cached selection
+    with zero overhead.
+
+    Args:
+        act_fp4: Activation tensor [m, k] in FP4 format (packed in uint8)
+        weight: Weight tensor [n, k] in FP4 format (packed in uint8)
+        act_sf: Activation scale factors
+        weight_scale: Weight scale factors
+        alpha: Scaling factor (as torch.Tensor for CUTLASS/cuBLASLt compatibility)
+        output_dtype: Output data type
+        to_userbuffers: Whether to use user buffers (CUTLASS/cuBLASLt only)
+        backend: Backend selection, one of:
+            - 'auto': AutoTuner automatically selects best backend (default)
+            - 'cutlass': Force use CUTLASS (FP4GemmRunner)
+            - 'cublaslt': Force use cuBLASLt (CublasLtFP4GemmRunner)
+            - 'cutedsl': Force use CuteDSL (CuteDSLNVFP4Wrapper)
+            - 'cuda_core': Force use CUDA Core (CudaCoreNVFP4Runner, requires SM >= 89, M <= 16)
+
+    Returns:
+        Output tensor [m, n] with dtype=output_dtype
+
+    Raises:
+        ValueError: If backend is invalid/unavailable
+    """
+
+    # Validate backend parameter
+    valid_backends = ['auto', 'cutlass', 'cublaslt', 'cutedsl', 'cuda_core']
+    if backend not in valid_backends:
+        raise ValueError(
+            f"Invalid backend '{backend}'. Must be one of {valid_backends}")
+
+    # Build list of runners based on backend parameter
+    runners = []
+
+    # CUDA Core runner can be enabled via backend='cuda_core' (not in auto mode by default)
+    # This avoids AutoTuner cache incompatibility issues
+    if backend == "cuda_core":
```

Comment:
This makes `cuda_core` backend mutually exclusive to `auto` backend? Sounds like a functionaly regression. Because we go from hueristically chosen cuda_core (if `module.enable_cuda_core` is true) or CUTLASS based on problem shape, to forcing the user to make choices between cuda_core and auto ahead of time.

Can you describe what the autotuner cache incompatibility issue is like?